### PR TITLE
complete the cv_bridge extension:

### PR DESCRIPTION
**Pull request by talregev**
complete the cv_bridge extension:
- Add CvCompressedImage class that convert from CompressedImage msg to opencv image and vice versa.
- Add two examples from [wiki](http://wiki.ros.org/cv_bridge/Tutorials/UsingCvBridgeToConvertBetweenROSImagesAndOpenCVImages):  
  That can show how CvImage and CvCompressedImage can be used.

### CODE REVIEW

**Comment by talregev** [comment]
@vrabaud
I have extend cv_bridge. please look at it.

**Comment by vrabaud** [comment]
ok, that is very valuable thx. I'll have a look at it. Please do not bump the version number.

**Comment by talregev** [comment]
ok.
should I commit a change that will roll back the version number?

thanks.

**Comment by vrabaud** [comment]
yes please. And don't hesitate to squash your commits so that each has a distinct purpose.

**Comment by talregev** [comment]
Roll back to version 1.11.8

**Comment by vrabaud** [comment]
it's overall ok but it's confusing at times: you create sections of code that you then delete.
Please squash  0b49528 and 1523f22
Please squash  fc5dfea and 6af29d3

**Comment by talregev** [comment]
How I squash commits?

**Comment by vrabaud** [comment]
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by talregev** [comment]
can you write me the command?
git rebase -s 0b49528  -s 1523f22

**Comment by vrabaud** [comment]
just follow the tutorials: `git rebase -i HEAD~10`, and then write s in the left column for the commits you want to squash with the above (you can copy / paste lines to move them around).

**Comment by talregev** [comment]
> that function seems like a lot of duplicated code, couldn't you call toCvCopyImpl from it ?

Meybe we can create toCvCopyImplHelper, that will bring the correct image?

**Comment by talregev** [comment]
I think we first review all the code, then we squash commits.
I will add toCvCopyImplHelper to avoid code duplication.

File: cv_bridge/src/cv_bridge.cpp [hunk 1]
```diff
@@ -393,4 +398,162 @@ CvImagePtr cvtColor(const CvImageConstPtr& source,
...
+// Internal, used by toCvCopy and cvtColor
+CvCompressedImagePtr toCvCompressedCopyImpl(const cv::Mat& source,
+                        const std_msgs::Header& src_header,
+                        const std::string& src_encoding,
+                        const std::string& dst_encoding)
+{
+  /// @todo Handle endianness - e.g. 16-bit dc1394 camera images are big-endian
+
+  // Copy metadata
+	std::cout << "toCvCompressedCopyImpl()" << std::endl;
```
Comment (vrabaud) [inline cv_bridge/src/cv_bridge.cpp]: please remove all std::cout

**Comment by vrabaud** [comment]
actually, I disagree with that pull request: your CvCompressedImage structure decompresses a message and stores it in a cv::Mat in an uncompressed way. The original goal of CvImage was to shared data between OpenCV and a cv::Mat for optimization.

In your case, data cannot be shared so there is no need for such a structure. On the other hand, sensor_msgs::CompressedImage needs to be supported. So I would suggest the following:
- create a toCompressedImageMsg(sensor_msgs::CompressedImage& ros_image, const std::string& dst_format = std::string()) member function in CvImage, that is basically your CvCompressedImage::toImageMsg
- have toCvCopy and cvtColor return a CvImagePtr
- delete the toCvShare you created

Feedback ?

**Comment by talregev** [comment]
> actually, I disagree with that pull request: your CvCompressedImage structure decompresses a message and stores it in a cv::Mat in an uncompressed way. The original goal of CvImage was to shared data between OpenCV and a cv::Mat for optimization.
> 
> In your case, data cannot be shared so there is no need for such a structure. On the other hand, sensor_msgs::CompressedImage needs to be supported. So I would suggest the following:
> 
> create a toCompressedImageMsg(sensor_msgs::CompressedImage& ros_image, const std::string& dst_format = std::string()) member function in CvImage, that is basically your CvCompressedImage::toImageMsg
> have toCvCopy and cvtColor return a CvImagePtr
> delete the toCvShare you created
> Feedback ?

Meybe at CvImage we have two methods:
- CompressedImage::toCompressedImageMsg()
- Image::toImageMsg()

And we have 4 additional methods:
- CvImageConstPtr toCvShare(const sensor_msgs::CompressedImageConstPtr& source,
                        const std::string& encoding = std::string());
- CvImageConstPtr toCvShare(const sensor_msgs::CompressedImage& source,
                        const boost::shared_ptr<void const>& tracked_object,
                        const std::string& encoding = std::string());  
  that will be friend of CvImage
-  CvImagePtr toCvCopy(const sensor_msgs::CompressedImageConstPtr& source,
                  const std::string& encoding = std::string());
- CvImagePtr toCvCopy(const sensor_msgs::CompressedImage& source,
                  const std::string& encoding = std::string());

Feedback?

**Comment by talregev** [comment]
what basically does toCvShare?

**Comment by vrabaud** [comment]
toCvShare fills a CvImage with the data from the message: data is shared between the two.

**Comment by talregev** [comment]
so we can drop toCvShare for ImageCompressed? 

#### Edit

But I can still implement toCvShare for  ImageCompressed,
for encoding, always fill: encoding = "bgr8".

Feedback?
Tell me if you think it ok, then I implement this.

**Comment by vrabaud** [comment]
well, there is nothing to share. You get a  ImageCompressed as an input, you decompress it and put it inside a CvImage. And that's it. Your original ImageCompressed can now have its own life.

toCvShare is simply for optimization: CvImage and Image do have a buffer to store the image. For optimization, toCvShare allows to have the same buffer shared between the two.

**Comment by talregev** [comment]
ok.

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 2]
```diff
@@ -256,10 +428,42 @@ template<> struct Definition<cv_bridge::CvImage>
 
 template<> struct HasHeader<cv_bridge::CvImage> : TrueType {};
 
+/////////////////////////////////////// CvCompressedImage ///////////////////////////////////////////
+
+template<> struct MD5Sum<cv_bridge::CvCompressedImage>
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: please keep that function

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 3]
```diff
@@ -319,18 +574,40 @@ template<> struct Printer<cv_bridge::CvImage>
   }
 };
 
+/////////////////////////////////////// CvCompressedImage ///////////////////////////////////////////
+
+template<> struct Printer<cv_bridge::CvCompressedImage>
+{
+  template<typename Stream>
+  static void stream(Stream& s, const std::string& indent, const cv_bridge::CvCompressedImage& m)
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: please implement

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 4]
```diff
@@ -306,10 +510,61 @@ template<> struct Serializer<cv_bridge::CvImage>
...
 };
 
+/////////////////////////////////////// CvCompressedImage ///////////////////////////////////////////
+
+template<> struct Serializer<cv_bridge::CvCompressedImage>
+{
+  /// @todo Still ignoring endianness...
+
+  template<typename Stream>
+  inline static void write(Stream& stream, const cv_bridge::CvCompressedImage& m)
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: please keep

File: cv_bridge/include/cv_bridge/cv_bridge.h (hunk 3 above)
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: oh, my bad, that was not done for Image, just leave as is for now.

File: cv_bridge/include/cv_bridge/cv_bridge.h (hunk 2 above)
Comment (talregev) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: > please keep that function

How to keep?
I am going to delete CvCompressedImage class.

File: cv_bridge/include/cv_bridge/cv_bridge.h (hunk 4 above)
Comment (talregev) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: > please keep

I am going to delete CvCompressedImage class.

File: cv_bridge/include/cv_bridge/cv_bridge.h (hunk 2 above)
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: yes, remove: the proper way to publish a compresed image should be through image transport anyway

File: cv_bridge/include/cv_bridge/cv_bridge.h (hunk 4 above)
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: same, sorry, please delete all that

**Comment by talregev** [comment]
Change API, and check that is working.
How I do squash commit?
`git rebase -i HEAD~10` should work?
or because I do one more commit it should be:
`git rebase -i HEAD~11`  ?

**Comment by vrabaud** [comment]
`git rebase -i HEAD~10` just opens up a text UI to ask you what to do with the last ten commits. In your case, you have less than 10 commits so just keep it as is

**Comment by talregev** [comment]
I do 
`git rebase -i HEAD~7` (because I have 7 new commit from you).
But it didn't change anything.

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 5]
```diff
@@ -229,6 +291,8 @@ namespace ros {
 
 namespace message_traits {
 
+/////////////////////////////////////// CvImage ///////////////////////////////////////////
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: no need for those comments anymore I guess

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 6]
```diff
@@ -93,6 +98,22 @@ class CvImage
    */
   void toImageMsg(sensor_msgs::Image& ros_image) const;
 
+  /**
+   * \brief Convert this message to a ROS sensor_msgs::Image message.
+   *
+   * The returned sensor_msgs::Image message contains a copy of the image data.
+   */
+  sensor_msgs::CompressedImagePtr toCompressedImageMsg() const;
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: please have it be like above `void toCompressedImageMsg(sensor_msgs::CompressedImagePtr& ros_image) const;`

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 7]
```diff
@@ -93,6 +98,22 @@ class CvImage
    */
   void toImageMsg(sensor_msgs::Image& ros_image) const;
 
+  /**
+   * \brief Convert this message to a ROS sensor_msgs::Image message.
+   *
+   * The returned sensor_msgs::Image message contains a copy of the image data.
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: fix doc

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 8]
```diff
@@ -93,6 +98,22 @@ class CvImage
...
+   */
+  sensor_msgs::CompressedImagePtr toCompressedImageMsg() const;
+
+  /**
+   * \brief Copy the message data to a ROS sensor_msgs::Image message.
+   *
+   * This overload is intended mainly for aggregate messages such as stereo_msgs::DisparityImage,
+   * which contains a sensor_msgs::Image as a data member.
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: fix doc

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 9]
```diff
@@ -200,6 +222,47 @@ CvImageConstPtr toCvShare(const sensor_msgs::Image& source,
                           const std::string& encoding = std::string());
 
 /**
+ * \brief Convert a sensor_msgs::Image message to an OpenCV-compatible CvImage, copying the
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: fix doc

Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h]: actually, no need to copy the doc, just declare it right before the non-compressed version

File: cv_bridge/src/cv_bridge.cpp [hunk 10]
```diff
@@ -393,4 +398,86 @@ CvImagePtr cvtColor(const CvImageConstPtr& source,
...
+  return ptr;
+}
+
+void CvImage::toCompressedImageMsg(sensor_msgs::CompressedImage& ros_image, const std::string& dst_format) const
+{
+  ros_image.header = header;
+  cv::Mat image;
+  if(encoding != enc::BGR8)
+  {
+	  CvImagePtr tempThis = boost::make_shared<CvImage>(*this);
```
Comment (vrabaud) [inline cv_bridge/src/cv_bridge.cpp]: please don't use tabs

**Comment by talregev** [comment]
done.

**Comment by talregev** [comment]
Now, if there is no more changes, I will squash commits.

File: cv_bridge/include/cv_bridge/cv_bridge.h [hunk 11]
```diff
@@ -93,6 +98,9 @@ class CvImage
    */
   void toImageMsg(sensor_msgs::Image& ros_image) const;
 
+  void toCompressedImageMsg(sensor_msgs::CompressedImage& ros_image, const std::string& dst_format = std::string()) const;
```
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h:115]: please explain the potential `dst_format`. Are those actually any standard ? Maybe an OpenCV reference.

File: opencv_apps/nodelet_plugins.xml [hunk 12]
```diff
@@ -63,6 +63,14 @@
   <class name="simple_flow/simple_flow" type="simple_flow::SimpleFlowNodelet" base_class_type="nodelet::Nodelet">
     <description>Nodelet of SimpleFlow optical flow algorithm</description>
   </class>
+
```
Comment (vrabaud) [inline opencv_apps/nodelet_plugins.xml:66]: inconsistent indentation in those lines.

**Comment by vrabaud** [comment]
ok, except for the last comments, I think it's good. Please fix and squash all that !

File: cv_bridge/include/cv_bridge/cv_bridge.h (hunk 11 above)
Comment (talregev) [inline cv_bridge/include/cv_bridge/cv_bridge.h:115]: it a string that you can compressed for format opencv has suport.
even in there code document they don't show OpenCv const or Reference. 
[opencv doc](http://docs.opencv.org/modules/highgui/doc/reading_and_writing_images_and_video.html#imencode):

**Comment by talregev** [comment]
You can fix the example if you want after the merge,
and maybe give them better names.

File: cv_bridge/include/cv_bridge/cv_bridge.h (hunk 11 above)
Comment (vrabaud) [inline cv_bridge/include/cv_bridge/cv_bridge.h:115]: sorry, I meant: explain as doxygen comments so that people using the function know how to use it.

**Comment by talregev** [comment]
Add a document for `dst_format`

**Comment by vrabaud** [comment]
Looks good ! Please squash everything.

**Comment by talregev** [comment]
Already Have :) :+1:

**Comment by talregev** [comment]
Final review?

**Comment by vrabaud** [comment]
Thx for all the work !

**Comment by talregev** [comment]
No problem :)
I updated cv_bridge for java accordingly.

**Pull request merged**